integration_test/tests/query_tests.rs

```diff
@@ -29,4 +29,70 @@ mod integration {
             _ => panic!("no data"),
         };
     }
+
+    fn columns_test(expected: &[[&str; 3]], filters: (&str, &str, &str)) {
+        use odbc_api::*;
+        let mut conn_string = crate::common::generate_default_connection_str();
+        conn_string.push_str("DATABASE=integration_test");
+
+        let env = Environment::new().unwrap();
+        let conn = env.connect_with_connection_string(&conn_string).unwrap();
+
+        let mut cursor = conn.columns(filters.0, "", filters.1, filters.2);
+        let mut i = 0;
+        while let Ok(Some(mut row)) = cursor.as_mut().unwrap().next_row() {
+            let mut buf = Vec::new();
+            let expected_row = expected[i];
+            i += 1;
+            row.get_text(1, &mut buf).unwrap();
+            assert_eq!(expected_row[0], std::str::from_utf8(&buf).unwrap());
+            row.get_text(3, &mut buf).unwrap();
+            assert_eq!(expected_row[1], std::str::from_utf8(&buf).unwrap());
+            row.get_text(4, &mut buf).unwrap();
+            assert_eq!(expected_row[2], std::str::from_utf8(&buf).unwrap());
+        }
+    }
+
+    #[test]
+    fn all_columns() {
+        let expected = [
+            ["integration_test", "example", "_id"],
+            ["integration_test", "example", "b"],
+            ["integration_test", "foo", "_id"],
+            ["integration_test", "foo", "a"],
+            ["integration_test_2", "example_2", "_id"],
+            ["integration_test_2", "example_2", "b"],
+        ];
+        columns_test(&expected, ("", "", ""));
+    }
+
+    #[test]
+    fn columns_with_column_filter() {
+        let expected = [
+            ["integration_test", "example", "_id"],
+            ["integration_test", "foo", "_id"],
+            ["integration_test_2", "example_2", "_id"],
+        ];
+        columns_test(&expected, ("", "", "%i%"));
+    }
+
+    #[test]
+    fn columns_with_collection_filter() {
+        let expected = [
+            ["integration_test", "example", "_id"],
+            ["integration_test", "example", "b"],
+            ["integration_test_2", "example_2", "_id"],
+            ["integration_test_2", "example_2", "b"],
+        ];
+        columns_test(&expected, ("", "%mp%", ""));
+    }
+
+    #[test]
+    fn columns_with_catalog() {
+        let expected = [
+            ["integration_test_2", "example_2", "_id"],
+            ["integration_test_2", "example_2", "b"],
+        ];
+        columns_test(&expected, ("integration_test_2", "", ""));
+    }
```

Comment:
Do you think there would be any value added by including a test that combines filters? Just by reading the code, I have no reason to think this would not work, so the independent filter tests provide sufficient coverage. Not only that, but also we might add more coverage for this and all other functions using the yaml test generator, like in the JDBC repo... The only reason I can think to add a test that has multiple filters is so we don't accidentally regress one day by altering the list_columns code. But still, that can be covered by a generated test.. What do you think -- worth adding `columns_test(&expected, ("integration_test_2", "example_2", "b")` or something like that, or is there no need at this time?

Reply:
I was thinking it wasn't necessary, but why not. But only one for a filter at every level not some power set :D